File: dotnet/src/LabelZoom.Sdk/LabelZoomClient.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Reflection;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using LabelZoom.Sdk.Conversion;

namespace LabelZoom.Sdk
{
    /// <summary>
    /// Client for the LabelZoom conversion API.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Thread-safe and intended to be long-lived. Create one per application, not one per request.
    /// </para>
    /// <para>
    /// <b>An API key is optional.</b> Constructed without one, the client uses the anonymous free
    /// tier: watermarked output, first label only, a 1 MB request cap, and no multi-page,
    /// JSON-target, or image-to-image conversion.
    /// </para>
    /// <example>
    /// <code>
    /// using var client = new LabelZoomClient();
    /// var result = await client.Convert()
    ///     .FromZpl("^XA^FO20,20^A0N,28^FDHello^FS^XZ")
    ///     .ToPng()
    ///     .WithDpi(300)
    ///     .ExecuteAsync();
    /// result.Save("label.png");
    /// </code>
    /// </example>
    /// </remarks>
    public sealed class LabelZoomClient : IDisposable
    {
        private const string RequestIdHeader = "X-LZ-Request-Id";
        private const int MaxMessageLength = 512;

        private static readonly string SdkVersion =
            typeof(LabelZoomClient).GetTypeInfo().Assembly.GetName().Version?.ToString(3) ?? "0.0.0";

        private readonly LabelZoomClientOptions _options;
        private readonly HttpClient _httpClient;
        private readonly bool _ownsHttpClient;
        private readonly string? _credential;
        private readonly string _userAgent;
        private readonly Random _jitter = new Random();

        private bool _disposed;

        /// <summary>Creates an anonymous client against the production API.</summary>
        public LabelZoomClient()
            : this(new LabelZoomClientOptions()) { }

        /// <summary>Creates a client authenticated with the given API key or JWT.</summary>
        /// <param name="apiKey">
        /// An <c>lz_live_</c>/<c>lz_test_</c> key or a JWT. Pass <c>null</c> to fall back to the
        /// <c>LABELZOOM_API_KEY</c> environment variable.
        /// </param>
        public LabelZoomClient(string? apiKey)
            : this(new LabelZoomClientOptions { ApiKey = apiKey }) { }

        /// <summary>Creates a client from an explicit options object.</summary>
        public LabelZoomClient(LabelZoomClientOptions options)
        {
            if (options is null)
            {
                throw new ArgumentNullException(nameof(options));
            }

            // Snapshot so later mutation of the caller's object cannot change this client.
            _options = options.Clone();
            _credential = _options.ResolveCredential();

            if (_options.HttpClient is not null)
            {
                _httpClient = _options.HttpClient;
                _ownsHttpClient = false;
            }
            else
            {
                _httpClient = _options.HttpMessageHandler is null
                    ? new HttpClient()
                    : new HttpClient(_options.HttpMessageHandler, disposeHandler: false);
                _ownsHttpClient = true;
            }

            if (_options.Timeout.HasValue && _ownsHttpClient)
            {
                _httpClient.Timeout = _options.Timeout.Value;
            }

            // The server parses a "LabelZoomStudio/" prefix as a Studio version and silently
            // changes PDF handling for versions <= 1.8.2, so the SDK's own token must come first.
            var userAgent = $"labelzoom-dotnet-sdk/{SdkVersion}";
            if (!string.IsNullOrWhiteSpace(_options.UserAgentSuffix))
            {
                userAgent += " " + _options.UserAgentSuffix!.Trim();
            }

            _userAgent = userAgent;
        }

        /// <summary>Whether a credential was resolved. False means anonymous free-tier requests.</summary>
        public bool IsAuthenticated => _credential is not null;

        /// <summary>Starts building a conversion.</summary>
        public ConversionRequestBuilder Convert()
        {
            ThrowIfDisposed();
            return new ConversionRequestBuilder(this);
        }

        internal async Task<ConversionResult> ExecuteAsync(
            SourceFormat source,
            TargetFormat target,
            byte[] body,
            string contentType,
            ConversionParameters parameters,
            CancellationToken cancellationToken)
        {
            ThrowIfDisposed();

            var uri = BuildUri(source, target, parameters);
            var attempts = _options.MaxRetries + 1;

            for (var attempt = 1; ; attempt++)
            {
                cancellationToken.ThrowIfCancellationRequested();

                HttpResponseMessage? response = null;
                try
                {
                    // A fresh request and content object per attempt: HttpRequestMessage cannot be
                    // reused, and a consumed HttpContent cannot be replayed.
                    using var request = new HttpRequestMessage(HttpMethod.Post, uri);
                    var content = new ByteArrayContent(body);
                    content.Headers.ContentType = MediaTypeHeaderValue.Parse(contentType);
                    request.Content = content;

                    // Accept must be */*. The server's `produces` list omits image/gif, image/bmp
                    // and image/jpeg, so naming the target's exact media type yields a 406 from
                    // content negotiation before the handler ever runs.
                    request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("*/*"));
                    request.Headers.TryAddWithoutValidation("User-Agent", _userAgent);

                    if (_credential is not null)
                    {
                        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _credential);
                    }

                    response = await _httpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);

                    if (response.IsSuccessStatusCode)
                    {
                        return await BuildResultAsync(response).ConfigureAwait(false);
                    }

                    var error = await BuildExceptionAsync(response).ConfigureAwait(false);

                    if (attempt >= attempts || !IsRetryable(response.StatusCode))
                    {
                        throw error;
                    }

                    var retryAfter = (error as LabelZoomRateLimitedException)?.RetryAfterSeconds;
                    await DelayAsync(attempt, retryAfter, cancellationToken).ConfigureAwait(false);
                }
                catch (HttpRequestException) when (attempt < attempts)
                {
                    // Transport-level failure: connection refused, DNS, TLS, dropped socket.
                    await DelayAsync(attempt, null, cancellationToken).ConfigureAwait(false);
                }
                catch (TaskCanceledException) when (attempt < attempts && !cancellationToken.IsCancellationRequested)
                {
                    // HttpClient surfaces its own timeout as TaskCanceledException. The guard
                    // distinguishes that from the caller actually cancelling.
                    await DelayAsync(attempt, null, cancellationToken).ConfigureAwait(false);
                }
                finally
                {
                    response?.Dispose();
                }
            }
        }

        internal Uri BuildUri(SourceFormat source, TargetFormat target, ConversionParameters parameters)
        {
            var path = $"{_options.BaseUrl}/api/v2/convert/{source.ToWireToken()}/to/{target.ToWireToken()}";

            var query = new List<string>();
            var json = parameters.ToJson();
            if (json is not null)
            {
                query.Add("params=" + Uri.EscapeDataString(json));
            }

            foreach (var pair in parameters.RawQueryParameters)
            {
                query.Add(Uri.EscapeDataString(pair.Key) + "=" + Uri.EscapeDataString(pair.Value));
            }

            // No options set means no query string at all, not an empty "?params={}".
            return new Uri(query.Count == 0 ? path : path + "?" + string.Join("&", query));
        }

        private async Task<ConversionResult> BuildResultAsync(HttpResponseMessage response)
        {
            var bytes = await response.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
            var contentType = response.Content.Headers.ContentType;

            Encoding encoding = Encoding.UTF8;
            if (!string.IsNullOrEmpty(contentType?.CharSet))
            {
                try
                {
                    encoding = Encoding.GetEncoding(contentType!.CharSet!.Trim('"'));
                }
                catch (ArgumentException)
                {
                    // An unrecognized charset is not worth failing an otherwise good conversion.
                    encoding = Encoding.UTF8;
                }
            }

            return new ConversionResult(
                bytes,
                contentType?.ToString(),
                (int)response.StatusCode,
                ReadRequestId(response),
                encoding);
        }

        private async Task<LabelZoomException> BuildExceptionAsync(HttpResponseMessage response)
        {
            string rawBody;
            try
            {
                rawBody = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
            }
            catch (Exception)
            {
                rawBody = string.Empty;
            }

            var message = ExtractMessage(rawBody, response.ReasonPhrase);
            var requestId = ReadRequestId(response);
            var status = response.StatusCode;

            return (int)status switch
            {
                400 => new LabelZoomBadRequestException(message, requestId, rawBody),
                401 => new LabelZoomUnauthorizedException(message, requestId, rawBody),
                403 => new LabelZoomForbiddenException(
                    message, requestId, rawBody, IsPaidFeatureMessage(message)),
                404 => new LabelZoomNotFoundException(message, requestId, rawBody),
                413 => new LabelZoomPayloadTooLargeException(message, requestId, rawBody),
                429 => new LabelZoomRateLimitedException(
                    message, requestId, rawBody, ReadRetryAfterSeconds(response)),
                >= 500 => new LabelZoomServerException(status, message, requestId, rawBody),
                _ => new LabelZoomException(status, message, requestId, rawBody),
            };
        }

        /// <summary>
        /// Pulls the human-readable detail out of an error body.
        /// </summary>
        /// <remarks>
        /// Both error shapes in play put it on <c>message</c>: the gateway returns
        /// <c>{"message": "..."}</c> and Spring returns
        /// <c>{"timestamp","status","error","message","path"}</c>. Anything else — a rate-limit body
        /// keyed on <c>error</c>, an HTML 502, a truncated JSON fragment — falls through to the raw
        /// text, then to the reason phrase.
        /// </remarks>
        internal static string ExtractMessage(string? rawBody, string? reasonPhrase)
        {
            if (!string.IsNullOrWhiteSpace(rawBody))
            {
                try
                {
                    using var document = JsonDocument.Parse(rawBody!);
                    if (document.RootElement.ValueKind == JsonValueKind.Object &&
                        document.RootElement.TryGetProperty("message", out var messageElement) &&
                        messageElement.ValueKind == JsonValueKind.String)
                    {
                        var parsed = messageElement.GetString();
                        if (!string.IsNullOrWhiteSpace(parsed))
                        {
                            return Truncate(parsed!);
                        }
                    }
                }
                catch (JsonException)
                {
                    // Not JSON, or malformed. The raw body is still the most useful thing we have.
                }

                return Truncate(rawBody!.Trim());
            }

            return string.IsNullOrWhiteSpace(reasonPhrase)
                ? "The LabelZoom API returned an error with no response body."
                : reasonPhrase!;
        }

        private static string Truncate(string value) =>
            value.Length <= MaxMessageLength ? value : value.Substring(0, MaxMessageLength);

        private static bool IsPaidFeatureMessage(string message) =>
            message.IndexOf("paid feature", StringComparison.OrdinalIgnoreCase) >= 0;

        private static bool IsRetryable(HttpStatusCode status) =>
            (int)status == 429 || (int)status >= 500;

        /// <summary>
        /// Reads <c>X-LZ-Request-Id</c> case-insensitively.
        /// </summary>
        /// <remarks>
        /// The gateway sets <c>X-LZ-Request-Id</c> but CORS-exposes it as <c>X-LZ-Request-ID</c>.
        /// <see cref="HttpHeaders"/> is already case-insensitive; the explicit fallback guards
        /// against a stack that is not.
        /// </remarks>
        private static string? ReadRequestId(HttpResponseMessage response)
        {
            if (response.Headers.TryGetValues(RequestIdHeader, out var values))
            {
                return values.FirstOrDefault();
            }

            foreach (var header in response.Headers)
            {
                if (string.Equals(header.Key, RequestIdHeader, StringComparison.OrdinalIgnoreCase))
                {
                    return header.Value.FirstOrDefault();
                }
            }

            return null;
        }

        private static int? ReadRetryAfterSeconds(HttpResponseMessage response)
        {
            var retryAfter = response.Headers.RetryAfter;
            if (retryAfter is null)
            {
                return null;
            }

            if (retryAfter.Delta.HasValue)
            {
                return (int)Math.Ceiling(retryAfter.Delta.Value.TotalSeconds);
            }

            if (retryAfter.Date.HasValue)
            {
                var seconds = (retryAfter.Date.Value - DateTimeOffset.UtcNow).TotalSeconds;
                return seconds > 0 ? (int)Math.Ceiling(seconds) : 0;
            }

            return null;
        }

        /// <summary>
        /// Waits before the next attempt: 1s, 2s, 4s, with full jitter, overridden by a longer
        /// <c>Retry-After</c>.
        /// </summary>
        private Task DelayAsync(int attempt, int? retryAfterSeconds, CancellationToken cancellationToken)
        {
            var backoffSeconds = Math.Pow(2, attempt - 1);
            var delay = TimeSpan.FromSeconds(backoffSeconds);

            if (_options.UseJitter)
            {
                double next;
                lock (_jitter)
                {
                    next = _jitter.NextDouble();
                }

                delay = TimeSpan.FromSeconds(backoffSeconds * next);
            }

            // The server knows better than the backoff curve when it tells us how long to wait.
            if (retryAfterSeconds.HasValue)
            {
                var serverDelay = TimeSpan.FromSeconds(retryAfterSeconds.Value);
                if (serverDelay > delay)
                {
                    delay = serverDelay;
                }
            }

            var sleep = _options.SleepAsync;
            return sleep is null ? Task.Delay(delay, cancellationToken) : sleep(delay, cancellationToken);
        }

        private void ThrowIfDisposed()
        {
            if (_disposed)
            {
                throw new ObjectDisposedException(nameof(LabelZoomClient));
            }
        }

        /// <summary>Disposes the internally created <see cref="HttpClient"/>, if any.</summary>
        public void Dispose()
        {
            if (_disposed)
            {
                return;
            }

            _disposed = true;

            if (_ownsHttpClient)
            {
                _httpClient.Dispose();
            }
        }
    }
}

File: java/src/main/java/com/labelzoom/sdk/LabelZoomClient.java
package com.labelzoom.sdk;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.TimeUnit;
import java.util.function.LongConsumer;

/**
 * Client for the LabelZoom conversion API.
 *
 * <p>Thread-safe and intended to be long-lived. Create one per application, not one per request.
 *
 * <p><b>An API key is optional.</b> Constructed without one, the client uses the anonymous free
 * tier: watermarked output, first label only, a 1 MB request cap, and no multi-page, JSON-target,
 * or image-to-image conversion.
 *
 * <pre>{@code
 * try (LabelZoomClient client = LabelZoomClient.builder().build()) {
 *     ConversionResult result = client.convert()
 *             .fromZpl("^XA^FO20,20^A0N,28^FDHello^FS^XZ")
 *             .toPng()
 *             .withDpi(300)
 *             .execute();
 *     result.save(Path.of("label.png"));
 * }
 * }</pre>
 */
public final class LabelZoomClient implements AutoCloseable {

    /** The production API host. */
    public static final String DEFAULT_BASE_URL = "https://api.labelzoom.com";

    /** The environment variable consulted when no credential is supplied. */
    public static final String API_KEY_ENVIRONMENT_VARIABLE = "LABELZOOM_API_KEY";

    private static final String REQUEST_ID_HEADER = "X-LZ-Request-Id";
    private static final int MAX_MESSAGE_LENGTH = 512;

    private final String baseUrl;
    private final String credential;
    private final int maxRetries;
    private final String userAgent;
    private final HttpTransport transport;
    private final boolean useJitter;
    private final LongConsumer sleeper;
    private final Random jitter = new Random();

    private LabelZoomClient(Builder builder) {
        this.baseUrl = builder.baseUrl.replaceAll("/+$", "");
        this.credential = resolveCredential(
                builder.apiKey, builder.apiKeySet, builder.environmentLookup);
        this.maxRetries = builder.maxRetries;
        this.useJitter = builder.useJitter;
        this.sleeper = builder.sleeper != null ? builder.sleeper : LabelZoomClient::sleepQuietly;

        String agent = "labelzoom-java-sdk/" + version() + " (java)";
        if (builder.userAgentSuffix != null && !builder.userAgentSuffix.isBlank()) {
            agent = agent + " " + builder.userAgentSuffix.trim();
        }
        this.userAgent = agent;

        this.transport = builder.transport != null
                ? builder.transport
                : new JdkHttpTransport(
                        HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build(),
                        builder.timeout);
    }

    /** Creates a builder. Every option has a working default. */
    public static Builder builder() {
        return new Builder();
    }

    /** Whether a credential was resolved. False means anonymous free-tier requests. */
    public boolean isAuthenticated() {
        return credential != null;
    }

    /** Starts building a conversion. */
    public ConversionRequestBuilder convert() {
        return new ConversionRequestBuilder(this);
    }

    /** Releases resources. The JDK transport holds none, so this is a no-op by default. */
    @Override
    public void close() {
        // Present so try-with-resources reads naturally and a future pooled transport can hook in.
    }

    ConversionResult execute(
            SourceFormat source,
            TargetFormat target,
            byte[] body,
            String contentType,
            Map<String, Object> params,
            Map<String, String> rawQuery) {

        String uri = buildUri(source, target, params, rawQuery);

        Map<String, String> headers = new LinkedHashMap<>();
        headers.put("Content-Type", contentType);
        // Accept must be */*. The server's `produces` list omits image/gif, image/bmp and
        // image/jpeg, so naming the target's exact media type yields a 406 from content
        // negotiation before the handler ever runs.
        headers.put("Accept", "*/*");
        headers.put("User-Agent", userAgent);
        if (credential != null) {
            headers.put("Authorization", "Bearer " + credential);
        }

        HttpTransport.Request request = new HttpTransport.Request("POST", uri, headers, body);
        int attempts = maxRetries + 1;

        for (int attempt = 1; ; attempt++) {
            HttpTransport.Response response;
            try {
                response = transport.send(request);
            } catch (IOException e) {
                if (attempt >= attempts) {
                    throw new UncheckedIOException("LabelZoom request failed", e);
                }
                delay(attempt, null);
                continue;
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new IllegalStateException("LabelZoom request was interrupted", e);
            }

            if (response.statusCode() >= 200 && response.statusCode() < 300) {
                return new ConversionResult(
                        response.body(),
                        response.header("Content-Type"),
                        response.statusCode(),
                        response.header(REQUEST_ID_HEADER));
            }

            LabelZoomException error = toException(response);

            if (attempt >= attempts || !isRetryable(response.statusCode())) {
                throw error;
            }

            // Read Retry-After from the response rather than the error: only the rate-limit type
            // exposes it, and RFC 9110 allows the header on 503 as well.
            delay(attempt, retryAfterSeconds(response));
        }
    }

    String buildUri(
            SourceFormat source,
            TargetFormat target,
            Map<String, Object> params,
            Map<String, String> rawQuery) {

        StringBuilder uri = new StringBuilder(baseUrl)
                .append("/api/v2/convert/")
                .append(source.wireToken())
                .append("/to/")
                .append(target.wireToken());

        StringBuilder query = new StringBuilder();
        if (!params.isEmpty()) {
            query.append("params=").append(encode(Json.write(params)));
        }
        for (Map.Entry<String, String> entry : rawQuery.entrySet()) {
            if (query.length() > 0) {
                query.append('&');
            }
            query.append(encode(entry.getKey())).append('=').append(encode(entry.getValue()));
        }

        // No options set means no query string at all, not an empty "?params={}".
        if (query.length() > 0) {
            uri.append('?').append(query);
        }
        return uri.toString();
    }

    private static String encode(String value) {
        // URLEncoder is form encoding, which differs from percent-encoding for space and a few
        // others. Correcting those keeps the JSON readable in server logs.
        return URLEncoder.encode(value, StandardCharsets.UTF_8)
                .replace("+", "%20")
                .replace("*", "%2A")
                .replace("%7E", "~");
    }

    private LabelZoomException toException(HttpTransport.Response response) {
        String rawBody = response.body() == null
                ? ""
                : new String(response.body(), StandardCharsets.UTF_8);
        String message = extractMessage(rawBody);
        String requestId = response.header(REQUEST_ID_HEADER);
        int status = response.statusCode();

        return switch (status) {
            case 400 -> new LabelZoomBadRequestException(message, requestId, rawBody);
            case 401 -> new LabelZoomUnauthorizedException(message, requestId, rawBody);
            case 403 -> new LabelZoomForbiddenException(
                    message, requestId, rawBody,
                    message.toLowerCase(Locale.ROOT).contains("paid feature"));
            case 404 -> new LabelZoomNotFoundException(message, requestId, rawBody);
            case 413 -> new LabelZoomPayloadTooLargeException(message, requestId, rawBody);
            case 429 -> new LabelZoomRateLimitedException(
                    message, requestId, rawBody, retryAfterSeconds(response));
            default -> status >= 500
                    ? new LabelZoomServerException(status, message, requestId, rawBody)
                    : new LabelZoomException(status, message, requestId, rawBody);
        };
    }

    /**
     * Pulls the human-readable detail out of an error body.
     *
     * <p>Both error shapes in play put it on {@code message}: the gateway returns
     * {@code {"message": "..."}} and Spring returns
     * {@code {timestamp,status,error,message,path}}. Anything else — a rate-limit body keyed on
     * {@code error}, an HTML 502, a truncated fragment — falls through to the raw text.
     */
    static String extractMessage(String rawBody) {
        if (rawBody != null && !rawBody.isBlank()) {
            String parsed = JsonMessage.extract(rawBody);
            if (parsed != null && !parsed.isBlank()) {
                return truncate(parsed);
            }
            return truncate(rawBody.trim());
        }
        return "The LabelZoom API returned an error with no response body.";
    }

    private static String truncate(String value) {
        return value.length() <= MAX_MESSAGE_LENGTH ? value : value.substring(0, MAX_MESSAGE_LENGTH);
    }

    private static boolean isRetryable(int status) {
        return status == 429 || status >= 500;
    }

    private static Integer retryAfterSeconds(HttpTransport.Response response) {
        String header = response.header("Retry-After");
        if (header == null) {
            return null;
        }
        try {
            return Integer.valueOf(header.trim());
        } catch (NumberFormatException e) {
            // The HTTP-date form is legal but rare here; treating it as absent falls back to the
            // backoff curve, which is safe.
            return null;
        }
    }

    /** 1s, 2s, 4s with full jitter, overridden by a longer {@code Retry-After}. */
    private void delay(int attempt, Integer retryAfterSeconds) {
        long backoffMillis = 1000L << (attempt - 1);
        long delayMillis = useJitter
                ? (long) (backoffMillis * jitter.nextDouble())
                : backoffMillis;

        if (retryAfterSeconds != null) {
            long serverMillis = TimeUnit.SECONDS.toMillis(retryAfterSeconds);
            if (serverMillis > delayMillis) {
                delayMillis = serverMillis;
            }
        }
        sleeper.accept(delayMillis);
    }

    private static void sleepQuietly(long millis) {
        try {
            Thread.sleep(millis);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("Interrupted while waiting to retry", e);
        }
    }

    private static String resolveCredential(
            String apiKey, boolean explicitlySet, java.util.function.UnaryOperator<String> environment) {
        if (!explicitlySet) {
            String fromEnvironment = environment.apply(API_KEY_ENVIRONMENT_VARIABLE);
            return fromEnvironment == null || fromEnvironment.isEmpty() ? null : fromEnvironment;
        }
        // An explicit null or empty forces anonymous and must not fall back to the environment.
        return apiKey == null || apiKey.isEmpty() ? null : apiKey;
    }

    private static String version() {
        String implementation = LabelZoomClient.class.getPackage().getImplementationVersion();
        return implementation != null ? implementation : "0.0.0-dev";
    }

    /** Builder for {@link LabelZoomClient}. */
    public static final class Builder {

        private String baseUrl = DEFAULT_BASE_URL;
        private String apiKey;
        private boolean apiKeySet;
        private int maxRetries = 2;
        private Duration timeout;
        private String userAgentSuffix;
        private HttpTransport transport;
        private boolean useJitter = true;
        private LongConsumer sleeper;
        // System.getenv is not mockable in Java, so the lookup itself is the seam. Package-private
        // rather than public: this exists for the conformance suite, not for callers.
        private java.util.function.UnaryOperator<String> environmentLookup = System::getenv;

        private Builder() {
        }

        /**
         * The bearer credential — an {@code lz_live_}/{@code lz_test_} key or a JWT.
         *
         * <p>Not calling this at all reads {@code LABELZOOM_API_KEY} from the environment. Calling
         * it with {@code null} or {@code ""} forces anonymous mode and suppresses that fallback.
         */
        public Builder apiKey(String apiKey) {
            this.apiKey = apiKey;
            this.apiKeySet = true;
            return this;
        }

        /** API base URL. A path prefix is preserved, so a reverse proxy works as expected. */
        public Builder baseUrl(String baseUrl) {
            if (baseUrl == null || baseUrl.isBlank()) {
                throw new IllegalArgumentException("Base URL cannot be null or blank.");
            }
            this.baseUrl = baseUrl;
            return this;
        }

        /** Retries after the initial attempt. Defaults to 2 (3 attempts total); 0 disables. */
        public Builder maxRetries(int maxRetries) {
            if (maxRetries < 0) {
                throw new IllegalArgumentException("maxRetries cannot be negative.");
            }
            this.maxRetries = maxRetries;
            return this;
        }

        /** Per-request timeout. */
        public Builder timeout(Duration timeout) {
            this.timeout = timeout;
            return this;
        }

        /** Appended to the SDK's own User-Agent, for identifying your application. */
        public Builder userAgentSuffix(String suffix) {
            this.userAgentSuffix = suffix;
            return this;
        }

        /** Replaces the HTTP layer. The seam the conformance suite and your own tests use. */
        public Builder transport(HttpTransport transport) {
            this.transport = transport;
            return this;
        }

        /** Whether to apply full jitter to retry backoff. Turn off for deterministic tests. */
        public Builder useJitter(boolean useJitter) {
            this.useJitter = useJitter;
            return this;
        }

        /** Replaces the delay between retries. Substitute a recording no-op in tests. */
        public Builder sleeper(LongConsumer sleeper) {
            this.sleeper = sleeper;
            return this;
        }

        /** Replaces the environment lookup used to resolve {@code LABELZOOM_API_KEY}. */
        Builder environmentLookup(java.util.function.UnaryOperator<String> environmentLookup) {
            this.environmentLookup = environmentLookup;
            return this;
        }

        /** Builds the client. */
        public LabelZoomClient build() {
            return new LabelZoomClient(this);
        }
    }
}
